README.md

```diff
@@ -0,0 +1,109 @@
+# 🛸Project Goals🛸
+
+These are snatched from the original project (_OP_), with some adjustments and amendments.
+Even though the list is ordered, they are not (yet) ordered by priority, but merely to make each point adressable, as in `🛸2`.
+
+1. provide an idiomatic rust implementation for Google APIs, which includes _type safety_ and native _async_ operations.
+1. first-class documentation with cross-links and complete code-examples
+1. support all API features, including downloads and resumable uploads
+1. Convenient CLIs are provided on top of the API for use in shell scripts
+1. API and CLI generation can be customized easily to overcome issues with your particular API
+   * **Byron thinks that** we cannot assume to get it right for all APIs from the start unless we actually test everything ourselves. Thus if we enable people to help themselves, it will help us further down the line.
+1. Built-in debugging and tracing allows to understand what's going on, and what's wrong.
+   * **Byron thinks that** providing full output of requests and responses when using the CLI really helped. However, this time around there might be some more logging, using `tracing` or `log` at least. Here once again it becomes interesting to see if different systems can be supported, to allow people to tailor their experience based on their needs. `cargo features` could come a long way.
+1. The code we generate defines the standard for interacting with Google services via Rust.
+   * Google uses these crates! They are that good!
+1. The code base is made for accepting PRs and making contributions easy
+   * To stay relevant, people must contribute.
+   * The original authors won't stay around forever (see [GitPython](https://github.com/gitpython-developers/GitPython))
+1. _safety and resilience_ are built-in, allowing you to create highly available tools on top of it. For example, you can trigger retries for all operations that may temporarily fail, e.g. due to network outage.
+   * **Byron thinks that** this could be naturally supported by the async ecosystem, and thus helps slim down the delegate.
+
+# Learning from the past
+
+Let's keep in mind what worked and what didn't.
+
+## 🌈What worked well in _OP_🌈
+
+1. **`make` to track dependencies and drive tooling**
+   * Having built-in help was nice, and one go-to location to drive everything around the project
+1. **Building big data models**
+   * It was very helpful to have all data available in a tree
+   * Merging structured data into even bigger trees helped to keep all data in easy to edit, human readable files, even with the option to pull in 'override files' to patch API descriptions as needed. The latter was used with the [Drive API](https://github.com/Byron/google-apis-rs/blob/master/etc/api/drive/v2/drive-api_overrides.yaml#L1), even though I would only add such capability on an as-needed basis.
+1. **Having an off-the-shelf template engine**
```

Comment:
The template engine to use is obviously a critical decision. I wasn't sure where I would end up when I started. On the one hand syn/quote seem like the natural choice when generating rust code since that's what almost every proc-macro uses (and for good reason). On the other hand proc-macros are not generating nearly as much code as these api bindings so I thought a more full featured template engine (like askama or yarte) may be better suited to the task and that quote may feel too limiting. I decided to just start with syn/quote and figured it would be pretty easy to transpose that into a more feature-full template if needed.

In my personal opinion it's worked really well. quote is not very feature rich, only allowing idents as the interpolation variables. This often leads to some code before the template capturing a more complex variable into its distilled form, but that also has the benefit of making sure all your logic is in the rust code and not the template itself. I haven't felt the need to reach for one of the other template engines yet, and the benefits of working with templates that understand the type of tokens is nice.

Reply:
Something I clearly had in mind is to fully separate between templating and model generation. From what I see, `syn`/`quote` naturally force you to do that.
However, I would take you saying these seem limited very seriously, and rather start looking for a template engine right away.

Having separate tools/sub-commands for these would make snapshot testing very straightforward. For example, `cargo run -- gen-model ./tests/fixtures/methods-with-foo.json > /tmp/model.json` generates desired output which is compared to a snapshot to protect against regression.

Similarly, that would be used as input and we can verify the we get the correct result, like in `cargo run -- gen-lib < /tmp/model.json`, which again can be compared with a snapshot, `cargo checked`, or what not.

This could look (and be driven by) something similar to [what I used in share-secrets-safely](https://travis-ci.org/share-secrets-safely/cli/jobs/475992924#L858).

By now, there are nice libraries for snapshot testing in cargo tests as well, and I would totally consider using these, going all Rust. For me, using `bash et al` usually was simpler.